### PR TITLE
feat(function): Support compare variant with other data types

### LINKED ISSUE #5428: Feature: auto cast variant to other data type in where condition
**Summary**

auto cast variant to other data type in where condition
```sql
create table t_variant(a variant);
insert into t_variant select parse_json('[1,2,3]');
select * from t_variant where a[0]=1;
```

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support compare variant with other data types

## Changelog

- New Feature
- Build/Testing/CI

## Related Issues

Fixes #5428

### CODE REVIEW

File: common/functions/src/scalars/comparisons/comparison.rs [hunk 1]
```diff
@@ -260,12 +294,18 @@ where
...
 {
     fn eval(&self, l: &ColumnWithField, r: &ColumnWithField) -> Result<BooleanColumn> {
-        scalar_binary_op(
-            l.column(),
-            r.column(),
-            self.func.clone(),
-            &mut EvalContext::default(),
-        )
+        let func_ctx = FunctionContext::default();
+        let lhs = if self.need_cast && !l.data_type().data_type_id().is_string() {
```
Comment (fkuner) [inline common/functions/src/scalars/comparisons/comparison.rs:298]: why needs to judge `is_string()`?

File: common/functions/src/scalars/comparisons/comparison.rs [hunk 2]
```diff
@@ -231,22 +259,28 @@ pub trait ComparisonImpl: Sync + Send + Clone + 'static {
...
+    need_cast: bool,
     func: F,
     _phantom: PhantomData<(L, R)>,
 }
 
 impl<L: Scalar, R: Scalar, F> ComparisonScalarImpl<L, R, F>
 where F: Fn(L::RefType<'_>, R::RefType<'_>, &mut EvalContext) -> bool
 {
-    pub fn new(func: F) -> Self {
+    pub fn new(least_supertype: DataTypeImpl, need_cast: bool, func: F) -> Self {
```
Comment (sundy-li) [inline common/functions/src/scalars/comparisons/comparison.rs:280]: why we need `need_cast`, if the least_supertype is equal to the input type, we need cast.

Comment (sundy-li) [inline common/functions/src/scalars/comparisons/comparison.rs:280]: why we need `need_cast`, if the least_supertype is equal to the input type, we need cast.

Comment (b41sh) [reply common/functions/src/scalars/comparisons/comparison.rs:280]: I see, do you mean remove `need_cast` argument and compare the input data type with `least_supertype`?

Comment (sundy-li) [reply common/functions/src/scalars/comparisons/comparison.rs]: Yes.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**